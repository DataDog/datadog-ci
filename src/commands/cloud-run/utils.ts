import type {IService, ServicesClient as IServicesClient} from './types'

import chalk from 'chalk'
import {GoogleAuth} from 'google-auth-library'
import {diff} from 'jest-diff'

import {withSpinner} from './renderer'

/**
 * Check if the user is authenticated with GCP.
 * @returns true if the user is authenticated, false otherwise
 */
export const checkAuthentication = async () => {
  const auth = new GoogleAuth()
  try {
    await auth.getApplicationDefault()

    return true
  } catch (_) {
    return false
  }
}

/**
 * Recursively sort object keys to ensure consistent ordering
 */
const sortObjectKeys = (obj: any): any => {
  if (!obj) {
    return obj
  }

  if (Array.isArray(obj)) {
    return obj.map(sortObjectKeys)
  }

  if (typeof obj === 'object') {
    const sorted: any = {}
    Object.keys(obj)
      .sort()
      .forEach((key) => {
        sorted[key] = sortObjectKeys(obj[key])
      })

    return sorted
  }

  return obj
}

/**
 * Obfuscate sensitive values in a line if it contains a key with "_KEY"
 */
const obfuscateSensitiveValues = (line: string): string => {
  // Match hex strings of 16, 32, or 64 characters (common API key/token lengths)
  return line
    .replace(/("[0-9a-fA-F]{16}"|"[0-9a-fA-F]{32}"|"[0-9a-fA-F]{64}")/g, '"***"')
    .replace(/('[0-9a-fA-F]{16}'|'[0-9a-fA-F]{32}'|'[0-9a-fA-F]{64}')/g, "'***'")
}

/**
 * Generate a git diff-style comparison between two configurations
 * TODO(@nhulston): update Lambda and AAS instrument to show this diff
 * @param original The original configuration object
 * @param updated The updated configuration object
 * @returns A formatted diff string with colors
 */
export const generateConfigDiff = (original: any, updated: any): string => {
  // Sort keys consistently before comparison
  const sortedOriginal = sortObjectKeys(original)
  const sortedUpdated = sortObjectKeys(updated)

  const originalJson = JSON.stringify(sortedOriginal, undefined, 2)
  const updatedJson = JSON.stringify(sortedUpdated, undefined, 2)

  const obfuscatedOriginal = originalJson.split('\n').map(obfuscateSensitiveValues).join('\n')
  const obfuscatedUpdated = updatedJson.split('\n').map(obfuscateSensitiveValues).join('\n')

  const configDiff = diff(obfuscatedOriginal, obfuscatedUpdated, {
    aColor: chalk.red,
    bColor: chalk.green,
    omitAnnotationLines: true,
  })
  if (!configDiff || configDiff.includes('no visual difference')) {
    return chalk.gray('No changes detected.')
  }

  return configDiff
}

export const fetchServiceConfigs = async (
  client: IServicesClient,
  project: string,
  region: string,
  services: string[]
) => {
  const existingServiceConfigs: IService[] = []
  for (const serviceName of services) {
    const servicePath = client.servicePath(project, region, serviceName)

    const existingService = await withSpinner(
      `Fetching configuration for ${chalk.bold(serviceName)}...`,
      async () => {
        try {
          const [serv] = await client.getService({name: servicePath})

          return serv
        } catch (error) {
          throw new Error(
            `Service ${serviceName} not found in project ${project}, region ${region}.\n\nNo services were instrumented.\n`
          )
        }
      },
      `Fetched service configuration for ${chalk.bold(serviceName)}`
    )
    existingServiceConfigs.push(existingService)
  }

  return existingServiceConfigs
}
