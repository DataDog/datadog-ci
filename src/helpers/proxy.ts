import * as http from 'http'
import * as https from 'https'

import {Agent, AgentConnectOpts} from 'agent-base'
import {HttpProxyAgent} from 'http-proxy-agent'
import {HttpsProxyAgent} from 'https-proxy-agent'
import {getProxyForUrl} from 'proxy-from-env'
import {SocksProxyAgent} from 'socks-proxy-agent'

type ProxyType = 'http' | 'https' | 'socks' | 'socks4' | 'socks4a' | 'socks5' | 'socks5h'

export interface ProxyConfiguration {
  auth?: {
    password: string
    username: string
  }
  host?: string
  port?: number
  protocol: ProxyType
}

export class ProxyAgent extends Agent {
  private httpAgent: http.Agent
  private httpsAgent: http.Agent

  constructor() {
    super()
    this.httpAgent = new http.Agent()
    this.httpsAgent = new https.Agent()
  }

  public async connect(req: http.ClientRequest, opts: AgentConnectOpts): Promise<http.Agent> {
    const {secureEndpoint} = opts

    const protocol = secureEndpoint ? 'https:' : 'http:'
    const host = req.getHeader('host')
    const url = new URL(req.path, `${protocol}//${host}`).href

    const proxyUrl = getProxyForUrl(url)
    if (proxyUrl) {
      const proxyAgent = getProxyAgentForUrl(proxyUrl)
      if (proxyAgent) {
        return proxyAgent
      }
    }

    return secureEndpoint ? this.httpsAgent : this.httpAgent
  }

  public destroy(): void {
    // for (const agent of this.cache.values()) {
    //   agent.destroy()
    // }
    super.destroy()
  }
}

export const getProxyUrlFromConfiguration = (options?: ProxyConfiguration): string => {
  if (!options) {
    return ''
  }

  const {auth, host, port, protocol} = options

  if (!host || !port) {
    return ''
  }

  const authFragment = auth ? `${auth.username}:${auth.password}@` : ''

  return `${protocol}://${authFragment}${host}:${port}`
}

export const getProxyAgent = (proxyOpts?: ProxyConfiguration): http.Agent | undefined => {
  const proxyUrlFromConfiguration = getProxyUrlFromConfiguration(proxyOpts)
  if (proxyUrlFromConfiguration) {
    return getProxyAgentForUrl(proxyUrlFromConfiguration)
  }

  return new ProxyAgent() as http.Agent
}

const getProxyAgentForUrl = (proxyUrl: string) => {
  const {protocol} = new URL(proxyUrl)
  if (protocol === 'http:') {
    return new HttpProxyAgent(proxyUrl)
  } else if (protocol === 'https:') {
    return new HttpsProxyAgent(proxyUrl)
  } else if (protocol.startsWith('socks')) {
    return new SocksProxyAgent(proxyUrl)
  }
}
