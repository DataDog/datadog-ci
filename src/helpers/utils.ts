import fs, {existsSync} from 'fs'
import {promisify} from 'util'

import type {SpanTag, SpanTags} from './interfaces'

import {AxiosRequestConfig, default as axios} from 'axios'
import {BaseContext, CommandClass, Cli} from 'clipanion'
import deepExtend from 'deep-extend'
import {ProxyAgent} from 'proxy-agent'

export const DEFAULT_CONFIG_PATHS = ['datadog-ci.json']

export const pick = <T extends Record<any, any>, K extends keyof T>(base: T, keys: K[]) => {
  const definedKeys = keys.filter((key) => !!base[key])
  const pickedObject: Partial<T> = {}

  for (const key of definedKeys) {
    pickedObject[key] = base[key]
  }

  return pickedObject
}

export const getConfig = async (configPath: string) => {
  try {
    const configFile = await promisify(fs.readFile)(configPath, 'utf-8')

    return JSON.parse(configFile)
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new Error('Config file is not correct JSON')
    }
  }
}

const resolveConfigPath = ({
  configPath,
  defaultConfigPaths,
}: {
  configPath?: string
  defaultConfigPaths?: string[]
}): string | undefined => {
  if (configPath) {
    if (existsSync(configPath)) {
      return configPath
    }
    throw new Error('Config file not found')
  }

  if (defaultConfigPaths) {
    for (const path of defaultConfigPaths) {
      if (existsSync(path)) {
        return path
      }
    }
  }

  return undefined
}

/**
 * Applies configurations in this order of priority:
 * environment > config file > base config
 */
export const resolveConfigFromFileAndEnvironment = async <
  T extends Record<string, unknown>,
  U extends Record<string, unknown>
>(
  baseConfig: T,
  environment: U,
  params: {
    configPath?: string
    defaultConfigPaths?: string[]
    configFromFileCallback?: (configFromFile: any) => void
  }
): Promise<T & U> => {
  const configFromFile = await resolveConfigFromFile(baseConfig, params)

  if (params.configFromFileCallback) {
    params.configFromFileCallback(configFromFile)
  }

  return deepExtend(configFromFile, removeUndefinedValues(environment))
}

export const resolveConfigFromFile = async <T>(
  baseConfig: T,
  params: {configPath?: string; defaultConfigPaths?: string[]}
): Promise<T> => {
  const resolvedConfigPath = resolveConfigPath(params)
  if (!resolvedConfigPath) {
    return baseConfig
  }
  const parsedConfig = await getConfig(resolvedConfigPath)

  return deepExtend(baseConfig, parsedConfig)
}

/**
 * @deprecated Use resolveConfigFromFile instead for better error management
 */
export const parseConfigFile = async <T>(baseConfig: T, configPath?: string): Promise<T> => {
  try {
    const resolvedConfigPath = configPath ?? 'datadog-ci.json'
    const parsedConfig = await getConfig(resolvedConfigPath)

    return deepExtend(baseConfig, parsedConfig)
  } catch (e) {
    if (e.code === 'ENOENT' && configPath) {
      throw new Error('Config file not found')
    }

    if (e instanceof SyntaxError) {
      throw new Error('Config file is not correct JSON')
    }
  }

  return baseConfig
}

type ProxyType =
  | 'http'
  | 'https'
  | 'socks'
  | 'socks4'
  | 'socks4a'
  | 'socks5'
  | 'socks5h'
  | 'pac+data'
  | 'pac+file'
  | 'pac+ftp'
  | 'pac+http'
  | 'pac+https'

export interface ProxyConfiguration {
  auth?: {
    password: string
    username: string
  }
  host?: string
  port?: number
  protocol: ProxyType
}

export const getProxyUrl = (options?: ProxyConfiguration): string => {
  if (!options) {
    return ''
  }

  const {auth, host, port, protocol} = options

  if (!host || !port) {
    return ''
  }

  const authFragment = auth ? `${auth.username}:${auth.password}@` : ''

  return `${protocol}://${authFragment}${host}:${port}`
}

export interface RequestOptions {
  apiKey: string
  appKey?: string
  baseUrl: string
  headers?: Map<string, string>
  overrideUrl?: string
  proxyOpts?: ProxyConfiguration
}

export const getRequestBuilder = (options: RequestOptions) => {
  const {apiKey, appKey, baseUrl, overrideUrl, proxyOpts} = options
  const overrideArgs = (args: AxiosRequestConfig) => {
    const newArguments = {
      ...args,
      headers: {
        'DD-API-KEY': apiKey,
        ...(appKey ? {'DD-APPLICATION-KEY': appKey} : {}),
        ...args.headers,
      },
    }

    if (overrideUrl !== undefined) {
      newArguments.url = overrideUrl
    }

    const proxyAgent = getProxyAgent(proxyOpts)
    if (proxyAgent) {
      newArguments.httpAgent = proxyAgent
      newArguments.httpsAgent = proxyAgent
    }

    if (options.headers !== undefined) {
      options.headers.forEach((value, key) => {
        newArguments.headers[key] = value
      })
    }

    return newArguments
  }

  const baseConfiguration: AxiosRequestConfig = {
    baseURL: baseUrl,
    // Disabling proxy in Axios config as it's not working properly
    // the passed httpAgent/httpsAgent are handling the proxy instead.
    proxy: false,
  }

  return (args: AxiosRequestConfig) => axios.create(baseConfiguration)(overrideArgs(args))
}

export const getProxyAgent = (proxyOpts?: ProxyConfiguration): ProxyAgent => {
  const proxyUrlFromConfiguration = getProxyUrl(proxyOpts)
  if (!proxyOpts || proxyUrlFromConfiguration === '') {
    // Let the default proxy agent discover environment variables.
    return new ProxyAgent()
  }

  return new ProxyAgent({
    getProxyForUrl: (url) => {
      // Do not proxy the WebSocket connections.
      if (url?.match(/^wss?:/)) {
        return ''
      }

      return proxyUrlFromConfiguration
    },
  })
}

export const getApiHostForSite = (site: string) => {
  switch (site) {
    case 'datad0g.com':
      return `app.${site}`
    case 'datadoghq.com':
    case 'datadoghq.eu':
    default:
      return `api.${site}`
  }
}

// The buildPath function is used to concatenate several paths. The goal is to have a function working for both unix
// paths and URL whereas standard path.join does not work with both.
export const buildPath = (...args: string[]) =>
  args
    .map((part, i) => {
      if (i === 0) {
        // For the first part, drop all / at the end of the path
        return part.trim().replace(/[\/]*$/g, '')
      } else {
        // For the following parts, remove all / at the beginning and at the end
        return part.trim().replace(/(^[\/]*|[\/]*$)/g, '')
      }
    })
    // Filter out empty parts
    .filter((x) => x.length)
    // Join all these parts with /
    .join('/')

export const removeEmptyValues = (tags: SpanTags) =>
  (Object.keys(tags) as SpanTag[]).reduce((filteredTags, tag) => {
    if (!tags[tag]) {
      return filteredTags
    }

    return {
      ...filteredTags,
      [tag]: tags[tag],
    }
  }, {})

export const removeUndefinedValues = <T extends {[key: string]: unknown}>(object: T): T => {
  const newObject = {...object}
  for (const [key, value] of Object.entries(newObject)) {
    if (value === undefined) {
      delete newObject[key]
    }
  }

  return newObject
}

export const normalizeRef = (ref: string | undefined) => {
  if (!ref) {
    return ref
  }

  return ref.replace(/origin\/|refs\/heads\/|tags\//gm, '')
}

export const pluralize = (nb: number, singular: string, plural: string) => {
  if (nb >= 2) {
    return `${nb} ${plural}`
  }

  return `${nb} ${singular}`
}

export const performSubCommand = (command: CommandClass<BaseContext>, commandArgs: string[], context: BaseContext) => {
  const cli = new Cli()
  cli.register(command)

  return cli.run(commandArgs, context)
}

export const filterSensitiveInfoFromRepository = (repositoryUrl: string | undefined) => {
  try {
    if (!repositoryUrl) {
      return repositoryUrl
    }
    if (repositoryUrl.startsWith('git@')) {
      return repositoryUrl
    }
    const {protocol, hostname, pathname} = new URL(repositoryUrl)
    if (!protocol || !hostname) {
      return repositoryUrl
    }

    return `${protocol}//${hostname}${pathname}`
  } catch (e) {
    return repositoryUrl
  }
}

// Removes sensitive info from the given git remote url and normalizes the url prefix.
// "git@github.com:" and "https://github.com/" prefixes will be normalized into "github.com/"
export const filterAndFormatGithubRemote = (rawRemote: string | undefined): string | undefined => {
  rawRemote = filterSensitiveInfoFromRepository(rawRemote)
  if (!rawRemote) {
    return rawRemote
  }
  rawRemote = rawRemote.replace(/git@github\.com:|https:\/\/github\.com\//, 'github.com/')

  return rawRemote
}

export const timedExecAsync = async <I, O>(f: (input: I) => Promise<O>, input: I): Promise<number> => {
  const initialTime = Date.now()
  await f(input)

  return (Date.now() - initialTime) / 1000
}

/**
 * Convert bytes to a formatted string in KB, MB, GB, etc.
 * Note: Lambda documentation uses MB (instead of Mib) to refer to 1024 KB, so we follow that style here
 * @param bytes
 * @param decimals
 */
export const formatBytes = (bytes: number, decimals = 2) => {
  if (!bytes) {
    return '0 Bytes'
  }

  if (bytes < 0) {
    throw Error("'bytes' can't be negative.")
  }

  const bytesPerKB = 1024
  const numDecimals = decimals < 0 ? 0 : decimals
  const units = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']
  const i = Math.floor(Math.log(bytes) / Math.log(bytesPerKB))
  const formattedBytes = parseFloat((bytes / Math.pow(bytesPerKB, i)).toFixed(numDecimals))

  return `${formattedBytes} ${units[i]}`
}

// Mask a string to hide sensitive values
export const maskString = (value: string) => {
  // Don't mask booleans
  if (value.toLowerCase() === 'true' || value.toLowerCase() === 'false') {
    return value
  }

  // Dont mask numbers
  if (!isNaN(Number(value))) {
    return value
  }

  // Mask entire string if it's short
  if (value.length < 12) {
    return '*'.repeat(16)
  }

  // Keep first two and last four characters if it's long
  return value.slice(0, 2) + '*'.repeat(10) + value.slice(-4)
}
