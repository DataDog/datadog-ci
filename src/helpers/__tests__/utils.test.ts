import http from 'http'
import {AddressInfo} from 'net'

import {AxiosPromise, AxiosRequestConfig, default as axios} from 'axios'
import {createProxy} from 'proxy'
import {ProxyAgent} from 'proxy-agent'

import * as ciUtils from '../utils'
import {formatBytes, maskString} from '../utils'

import {MOCK_DATADOG_API_KEY} from './fixtures'

describe('utils', () => {
  beforeEach(() => {
    jest.restoreAllMocks()
  })

  test('pick', () => {
    const initialHash = {a: 1, b: 2}

    let resultHash = ciUtils.pick(initialHash, ['a'])
    expect(Object.keys(resultHash).indexOf('b')).toBe(-1)
    expect(resultHash.a).toBe(1)

    resultHash = ciUtils.pick(initialHash, ['c'] as any)
    expect(Object.keys(resultHash).length).toBe(0)
  })

  describe('resolveConfigFromFile', () => {
    test('should read a config file', async () => {
      const config: any = await ciUtils.resolveConfigFromFile(
        {},
        {configPath: 'src/helpers/__tests__/config-file-fixtures/dummy-config-file.json'}
      )
      expect(config.configKey).toBe('newconfigvalue')
    })

    test('should throw an error if path is provided and config file is not found', async () => {
      const config = ciUtils.resolveConfigFromFile({}, {configPath: '/veryuniqueandabsentfile'})

      await expect(config).rejects.toEqual(Error('Config file not found'))
    })

    test('should have no effect if no config path is provided and default file is absent', async () => {
      const originalConfig = {}
      const config = await ciUtils.resolveConfigFromFile(originalConfig, {
        defaultConfigPaths: ['/veryuniqueandabsentfile'],
      })

      expect(config).toEqual(originalConfig)
    })

    test('should have no effect if no config path nor default path are provided', async () => {
      const originalConfig = {}
      const config = await ciUtils.resolveConfigFromFile(originalConfig, {})

      expect(config).toEqual(originalConfig)
    })

    test('should throw an error if JSON parsing fails', async () => {
      await expect(
        ciUtils.resolveConfigFromFile({}, {configPath: 'src/helpers/__tests__/config-file-fixtures/bad-json.json'})
      ).rejects.toEqual(Error('Config file is not correct JSON'))
    })

    test('config file should overwrite default configuration', async () => {
      const config: any = await ciUtils.resolveConfigFromFile(
        {configKey: 'oldValue'},
        {configPath: 'src/helpers/__tests__/config-file-fixtures/dummy-config-file.json'}
      )
      expect(config.configKey).toBe('newconfigvalue')
    })
  })

  describe('getRequestBuilder', () => {
    const fakeEndpointBuilder = (request: (args: AxiosRequestConfig) => AxiosPromise) => async () => request({})

    test('should add api key header', async () => {
      jest.spyOn(axios, 'create').mockImplementation((() => (args: AxiosRequestConfig) => args.headers) as any)
      const requestOptions = {
        apiKey: 'apiKey',
        baseUrl: 'http://fake-base.url/',
      }
      const request = ciUtils.getRequestBuilder(requestOptions)
      const fakeEndpoint = fakeEndpointBuilder(request)
      expect(await fakeEndpoint()).toStrictEqual({'DD-API-KEY': 'apiKey'})
    })

    test('should add api and application key header', async () => {
      jest.spyOn(axios, 'create').mockImplementation((() => (args: AxiosRequestConfig) => args.headers) as any)
      const requestOptions = {
        apiKey: 'apiKey',
        appKey: 'applicationKey',
        baseUrl: 'http://fake-base.url/',
      }
      const request = ciUtils.getRequestBuilder(requestOptions)
      const fakeEndpoint = fakeEndpointBuilder(request)
      expect(await fakeEndpoint()).toStrictEqual({'DD-API-KEY': 'apiKey', 'DD-APPLICATION-KEY': 'applicationKey'})
    })

    describe('proxy configuration', () => {
      test('should have a ProxyAgent by default', async () => {
        jest.spyOn(axios, 'create').mockImplementation((() => (args: AxiosRequestConfig) => args.httpsAgent) as any)
        const requestOptions = {
          apiKey: 'apiKey',
          appKey: 'applicationKey',
          baseUrl: 'http://fake-base.url/',
        }
        const request = ciUtils.getRequestBuilder(requestOptions)
        const fakeEndpoint = fakeEndpointBuilder(request)
        const httpsAgent = await fakeEndpoint()
        expect(httpsAgent).toBeDefined()
        expect(httpsAgent).toBeInstanceOf(ProxyAgent)
      })

      test('should add proxy configuration when explicitly defined', async () => {
        jest.spyOn(axios, 'create').mockImplementation((() => (args: AxiosRequestConfig) => args.httpsAgent) as any)
        const proxyOpts: ciUtils.ProxyConfiguration = {protocol: 'http', host: '1.2.3.4', port: 1234}
        const requestOptions = {
          apiKey: 'apiKey',
          appKey: 'applicationKey',
          baseUrl: 'http://fake-base.url/',
          proxyOpts,
        }
        const request = ciUtils.getRequestBuilder(requestOptions)
        const fakeEndpoint = fakeEndpointBuilder(request)
        const httpsAgent = await fakeEndpoint()
        expect(httpsAgent).toBeDefined()
        expect((httpsAgent as any).getProxyForUrl()).toBe('http://1.2.3.4:1234')
      })
    })

    test('should accept overrideUrl', async () => {
      jest.spyOn(axios, 'create').mockImplementation((() => (args: AxiosRequestConfig) => args.url) as any)
      const requestOptions = {
        apiKey: 'apiKey',
        appKey: 'applicationKey',
        baseUrl: 'http://fake-base.url/',
        overrideUrl: 'override/url',
      }
      const request = ciUtils.getRequestBuilder(requestOptions)
      const fakeEndpoint = fakeEndpointBuilder(request)
      expect(await fakeEndpoint()).toStrictEqual('override/url')
    })

    test('should accept additional headers', async () => {
      jest.spyOn(axios, 'create').mockImplementation((() => (args: AxiosRequestConfig) => args.headers) as any)
      const requestOptions = {
        apiKey: 'apiKey',
        appKey: 'applicationKey',
        baseUrl: 'http://fake-base.url/',
        headers: new Map([
          ['HEADER1', 'value1'],
          ['HEADER2', 'value2'],
        ]),
        overrideUrl: 'override/url',
      }
      const request = ciUtils.getRequestBuilder(requestOptions)
      const fakeEndpoint = fakeEndpointBuilder(request)
      expect(await fakeEndpoint()).toStrictEqual({
        'DD-API-KEY': 'apiKey',
        'DD-APPLICATION-KEY': 'applicationKey',
        HEADER1: 'value1',
        HEADER2: 'value2',
      })
    })
  })

  describe('getApiHostForSite', () => {
    it.each([
      ['datad0g.com', 'app.datad0g.com'],
      ['datadoghq.com', 'api.datadoghq.com'],
      ['datadoghq.eu', 'api.datadoghq.eu'],
      ['whitelabel.com', 'api.whitelabel.com'],
    ])('for site = %p, returns api host = %p', (site, expectedApiHost) => {
      expect(ciUtils.getApiHostForSite(site)).toEqual(expectedApiHost)
    })
  })

  describe('buildPath', () => {
    test('should return correct path', () => {
      const pathWithNoTrailingSlash = 'sourcemaps/js'
      const pathWithTrailingSlash = 'sourcemaps/js/'
      const fileName = 'file1.min.js'

      expect(ciUtils.buildPath(pathWithNoTrailingSlash, fileName)).toBe('sourcemaps/js/file1.min.js')
      expect(ciUtils.buildPath(pathWithTrailingSlash, fileName)).toBe('sourcemaps/js/file1.min.js')
    })
  })

  describe('getProxyUrl', () => {
    test('should return correct proxy URI', () => {
      expect(ciUtils.getProxyUrl()).toBe('')
      expect(ciUtils.getProxyUrl({protocol: 'http'})).toBe('')
      expect(ciUtils.getProxyUrl({host: '127.0.0.1', protocol: 'http'})).toBe('')
      expect(ciUtils.getProxyUrl({host: '127.0.0.1', port: 1234, protocol: 'http'})).toBe('http://127.0.0.1:1234')

      const auth = {password: 'pwd', username: 'john'}
      expect(ciUtils.getProxyUrl({auth, host: '127.0.0.1', port: 1234, protocol: 'http'})).toBe(
        'http://john:pwd@127.0.0.1:1234'
      )
      expect(ciUtils.getProxyUrl({auth, protocol: 'http'})).toBe('')
    })
  })

  // Test the different possibilities of proxy configuration of getRequestHelper.
  // All the calls to getRequestHelpers should be https calls, but to keep the test suite
  // simple tests are using http calls (testing with https would require us to add tls certs
  // and configure axios to trust these tls certs, which requires an agent config, which
  // interferes a bit with how the proxies are configured since they are configured through an
  // agent themselves.
  // Proxy of https requests is still tested in the proxy-agent library itself.
  describe('Proxy configuration', () => {
    let initialHttpProxyEnv: string | undefined

    beforeAll(() => {
      initialHttpProxyEnv = process.env.HTTP_PROXY
    })

    afterAll(() => {
      if (initialHttpProxyEnv !== undefined) {
        process.env.HTTP_PROXY = initialHttpProxyEnv
      } else {
        delete process.env.HTTP_PROXY
      }
    })

    beforeEach(() => {
      delete process.env.HTTP_PROXY
    })

    // Start a target http server and a proxy server listening on localhost,
    // returns the ports they listen to, a spy method allowing us to check if they've been
    // handling any requests, and a function to close them.
    const setupServer = async () => {
      // Create target http server
      const spyTargetServer = jest.fn()
      const targetHttpServer = http.createServer((_, res) => {
        spyTargetServer()
        res.end('response from target http server')
      })
      await new Promise<void>((resolve, reject) => {
        targetHttpServer.listen().once('listening', resolve).once('error', reject)
      })

      // Create proxy
      const proxyHttpServer = http.createServer()
      const proxyServer = createProxy(proxyHttpServer)
      const spyProxyServer = jest.fn()
      proxyHttpServer.on('request', spyProxyServer)
      await new Promise<void>((resolve, reject) => {
        proxyHttpServer.listen().once('listening', resolve).once('error', reject)
      })

      return {
        proxyServer: {
          close: async () =>
            new Promise<void>((resolve, reject) => {
              proxyServer.close((err) => {
                if (err) {
                  reject(err)
                }
                resolve()
              })
            }),
          port: (proxyHttpServer.address() as AddressInfo).port,
          spy: spyProxyServer,
        },
        targetServer: {
          close: async () =>
            new Promise<void>((resolve, reject) => {
              targetHttpServer.close((err: Error | undefined) => {
                if (err) {
                  reject(err)
                }
                resolve()
              })
            }),
          port: (targetHttpServer.address() as AddressInfo).port,
          spy: spyTargetServer,
        },
      }
    }

    test('Work without a proxy defined', async () => {
      const {proxyServer, targetServer} = await setupServer()
      try {
        const requestBuilder = ciUtils.getRequestBuilder({
          apiKey: 'abc',
          baseUrl: `http://localhost:${targetServer.port}`,
        })
        await requestBuilder({
          method: 'GET',
          url: 'test-from-proxy',
        })
        expect(targetServer.spy.mock.calls.length).toBe(1)
        expect(proxyServer.spy.mock.calls.length).toBe(0)
      } finally {
        await targetServer.close()
        await proxyServer.close()
      }
    })

    test('Proxy configured explicitly', async () => {
      const {proxyServer, targetServer} = await setupServer()
      try {
        const requestBuilder = ciUtils.getRequestBuilder({
          apiKey: 'abc',
          baseUrl: `http://localhost:${targetServer.port}`,
          proxyOpts: {
            host: 'localhost',
            port: proxyServer.port,
            protocol: 'http',
          },
        })
        await requestBuilder({
          method: 'GET',
          url: 'test-from-proxy',
        })
        expect(targetServer.spy.mock.calls.length).toBe(1)
        expect(proxyServer.spy.mock.calls.length).toBe(1)
      } finally {
        await targetServer.close()
        await proxyServer.close()
      }
    })

    test('Proxy configured through env var', async () => {
      const {proxyServer, targetServer} = await setupServer()
      try {
        process.env.HTTP_PROXY = `http://localhost:${proxyServer.port}`
        const requestBuilder = ciUtils.getRequestBuilder({
          apiKey: 'abc',
          baseUrl: `http://localhost:${targetServer.port}`,
        })
        await requestBuilder({
          method: 'GET',
          url: 'test-from-proxy',
        })
        expect(targetServer.spy.mock.calls.length).toBe(1)
        expect(proxyServer.spy.mock.calls.length).toBe(1)
      } finally {
        await targetServer.close()
        await proxyServer.close()
      }
    })

    test('Proxy configured explicitly takes precedence over env var', async () => {
      const {proxyServer, targetServer} = await setupServer()
      try {
        process.env.HTTP_PROXY = `http://incorrecthost:${proxyServer.port}`
        const requestBuilder = ciUtils.getRequestBuilder({
          apiKey: 'abc',
          baseUrl: `http://localhost:${targetServer.port}`,
          proxyOpts: {
            host: 'localhost',
            port: proxyServer.port,
            protocol: 'http',
          },
        })
        await requestBuilder({
          method: 'GET',
          url: 'test-from-proxy',
        })
        expect(targetServer.spy.mock.calls.length).toBe(1)
        expect(proxyServer.spy.mock.calls.length).toBe(1)
      } finally {
        await targetServer.close()
        await proxyServer.close()
      }
    })
  })

  describe('filterAndFormatGithubRemote', () => {
    test('git remotes get formatted correctly', async () => {
      expect(ciUtils.filterAndFormatGithubRemote('https://github.com/datadog/test.git')).toEqual(
        'github.com/datadog/test.git'
      )
      expect(ciUtils.filterAndFormatGithubRemote('git@github.com:datadog/test.git')).toEqual(
        'github.com/datadog/test.git'
      )
      expect(ciUtils.filterAndFormatGithubRemote('github.com/datadog/test.git')).toEqual('github.com/datadog/test.git')
    })
  })

  describe('formatBytes', () => {
    it('returns "0 Bytes" when input is 0', () => {
      expect(formatBytes(0)).toEqual('0 Bytes')
    })

    it('returns correct format for input in Bytes', () => {
      expect(formatBytes(500)).toEqual('500 Bytes')
    })

    it('returns correct format for input in KB', () => {
      expect(formatBytes(1024)).toEqual('1 KB')
      expect(formatBytes(1500, 2)).toEqual('1.46 KB')
    })

    it('returns correct format for input in MB', () => {
      expect(formatBytes(1048576)).toEqual('1 MB')
      expect(formatBytes(1572864, 2)).toEqual('1.5 MB')
    })

    it('respects the decimal parameter and rounds up when needed', () => {
      expect(formatBytes(2313561)).toEqual('2.21 MB')
      expect(formatBytes(2313561, 0)).toEqual('2 MB')
      expect(formatBytes(2313561, 1)).toEqual('2.2 MB')
      expect(formatBytes(2313561, 3)).toEqual('2.206 MB')
    })

    it('handles negative decimals by treating them as zero', () => {
      expect(formatBytes(1572864, -1)).toEqual('2 MB')
    })

    it('throws an error if the input is negative', () => {
      expect(() => formatBytes(-1000)).toThrow()
    })
  })

  describe('maskString', () => {
    it('should make the entire string if its length is less than 12', () => {
      expect(maskString('shortString')).toEqual('****************')
    })

    it('should keep the first two and last four characters for strings longer than 12 characters', () => {
      const original = 'abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz'
      const masked = 'ab**********wxyz'
      expect(maskString(original)).toEqual(masked)
    })

    it('should return empty string if input is empty', () => {
      expect(maskString('')).toEqual('')
    })

    it('should not mask booleans', () => {
      expect(maskString('true')).toEqual('true')
      expect(maskString('TrUe')).toEqual('TrUe')
      expect(maskString('false')).toEqual('false')
      expect(maskString('FALSE')).toEqual('FALSE')
      expect(maskString('trueee')).toEqual('****************')
    })

    it('should mask API keys correctly', () => {
      expect(maskString(MOCK_DATADOG_API_KEY)).toEqual('02**********33bd')
    })
  })
})

test('removeUndefinedValues', () => {
  // eslint-disable-next-line no-null/no-null
  expect(ciUtils.removeUndefinedValues({a: 'b', c: 'd', e: undefined, g: null})).toEqual({a: 'b', c: 'd', g: null})
})
