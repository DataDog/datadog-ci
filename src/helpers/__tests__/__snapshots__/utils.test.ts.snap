// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils maskConfig cloudrun config: should mask API key but not whitelisted environment variables 1`] = `
Object {
  "template": Object {
    "containers": Array [
      Object {
        "env": Array [
          Object {
            "name": "DD_API_KEY",
            "value": "02**********33bd",
            "values": "value",
          },
          Object {
            "name": "DD_TRACE_ENABLED",
            "value": "true",
            "values": "value",
          },
          Object {
            "name": "DD_SITE",
            "value": "datad0g.com",
            "values": "value",
          },
        ],
        "image": "gcr.io/datadog-sandbox/nicholas-hulston-docker-test",
      },
    ],
    "someData": "data",
  },
}
`;

exports[`utils maskConfig lambda config: should mask API key but not whitelisted environment variables 1`] = `
Object {
  "Environment": Object {
    "Variables": Object {
      "DD_API_KEY": "02**********33bd",
      "DD_LOG_LEVEL": "debug",
      "DD_SITE": "datadoghq.com",
    },
  },
  "FunctionArn": "arn:aws:lambda:us-east-1:123456789012:function:some-function",
  "FunctionName": "some-function",
}
`;

exports[`utils maskConfig should throw an error if the config is unknown 1`] = `"Unknown config format"`;
